File: typescript/Structural/11-Flyweight/Context/JobCard.ts
import { JobCategoryMetadata } from "../Flyweight/JobCategoryMetadata";
import { JobCategoryFactory } from "../Flyweight/JobCategoryFactory";

const shortId = () => Math.random().toString(36).slice(2, 10);

export class JobCard {
  readonly jobId: string = shortId();
  status: string = "Active";

  private readonly metadata: JobCategoryMetadata;

  constructor(
    readonly clientRef: string,
    readonly scheduledAt: Date,
    category: string,
  ) {
    this.metadata = JobCategoryFactory.getMetadata(category);
  }

  printSummary(): void {
    console.log(
      `\n  Job #${this.jobId} | Client: ${this.clientRef} | ${this.scheduledAt.toDateString()} | ${this.status}`,
    );
    this.metadata.printMetadata();
  }

  getMetadata(): JobCategoryMetadata {
    return this.metadata;
  }
}

File: csharp/Structural/11-Flyweight/Context/JobCard.cs
public class JobCard
{
    public string JobId { get; } = Guid.NewGuid().ToString()[..8];
    public string ClientRef { get; }
    public DateTime ScheduledAt { get; }
    public string Status { get; set; } = "Active";

    private readonly JobCategoryMetadata _metadata;

    public JobCard(string clientRef, DateTime scheduledAt, string category)
    {
        ClientRef = clientRef;
        ScheduledAt = scheduledAt;
        _metadata = JobCategoryFactory.GetMetadata(category);
    }

    public void PrintSummary()
    {
        Console.WriteLine($"\n  Job #{JobId} | Client: {ClientRef} | {ScheduledAt:dd MMM yyyy} | {Status}");
        _metadata.PrintMetadata();
    }
}
